typescript/sdk/src/providers/transactions/submitter/IcaTxSubmitter.ts

```diff
@@ -0,0 +1,203 @@
+  originInterchainAccountRouter: true,
+});
+
+/* eslint-disable-next-line @typescript-eslint/no-unused-vars */
+const EvmIcaTxSubmitterConstructorConfigSchema =
+  EvmIcaTxSubmitterConfigSchema.omit({ internalSubmitter: true }).required({
+    owner: true,
+  });
+
+type EvmIcaTxSubmitterConfig = z.infer<typeof EvmIcaTxSubmitterConfigSchema>;
+type EvmIcaTxSubmitterConstructorConfig = z.infer<
+  typeof EvmIcaTxSubmitterConstructorConfigSchema
+>;
+
+type SubmitterFactoryMapping<
+  E extends TxSubmitterType,
+  TConfig extends { type: E },
+  TResult,
+> = {
+  [K in E]: (
+    config: Extract<TConfig, { type: K }>,
+  ) => Promise<TResult> | TResult;
+};
+
+async function getInternalSubmitter(
+  chain: ChainName,
+  multiProvider: MultiProvider,
+  config: EvmIcaTxSubmitterConfig['internalSubmitter'],
+): Promise<TxSubmitterInterface<ProtocolType.Ethereum>> {
+  const internalSubmitterMap: SubmitterFactoryMapping<
+    EvmIcaTxSubmitterConfig['internalSubmitter']['type'],
+    EvmIcaTxSubmitterConfig['internalSubmitter'],
+    TxSubmitterInterface<ProtocolType.Ethereum>
+  > = {
+    [TxSubmitterType.GNOSIS_SAFE]: (config) => {
+      return EV5GnosisSafeTxSubmitter.create(multiProvider, {
+        chain,
+        ...config,
+      });
+    },
+    [TxSubmitterType.GNOSIS_TX_BUILDER]: (config) => {
+      return EV5GnosisSafeTxBuilder.create(multiProvider, {
+        chain,
+        ...config,
+      });
+    },
+    [TxSubmitterType.IMPERSONATED_ACCOUNT]: (config) => {
+      return new EV5ImpersonatedAccountTxSubmitter(multiProvider, {
+        chain,
+        ...config,
+      });
+    },
+    [TxSubmitterType.JSON_RPC]: (config) => {
+      return new EV5JsonRpcTxSubmitter(multiProvider, {
+        chain,
+        ...config,
+      });
+    },
+  };
+
+  const internalSubmitterFactory = internalSubmitterMap[config.type];
+  // Sanity check
+  if (!internalSubmitterFactory) {
+    throw new Error(
+      `Internal submitter factory not found for type: ${config.type}`,
+    );
+  }
+
+  return internalSubmitterFactory(config as any);
+}
+
+export class EvmIcaTxSubmitter
+  implements TxSubmitterInterface<ProtocolType.Ethereum>
+{
+  readonly txSubmitterType: TxSubmitterType =
+    TxSubmitterType.INTERCHAIN_ACCOUNT;
+
+  private constructor(
+    private readonly config: EvmIcaTxSubmitterConstructorConfig,
+    private readonly submitter: TxSubmitterInterface<ProtocolType.Ethereum>,
+    private readonly multiProvider: MultiProvider,
+    private readonly interchainAccountApp: InterchainAccount,
+  ) {}
+
+  static async fromConfig(
+    config: EvmIcaTxSubmitterConfig,
+    multiProvider: MultiProvider,
+  ): Promise<EvmIcaTxSubmitter> {
+    const internalSubmitter = await getInternalSubmitter(
+      config.chain,
+      multiProvider,
+      config.internalSubmitter,
+    );
+
+    const owner =
+      config.owner ?? (await multiProvider.getSignerAddress(config.chain));
```

Comment:
this seems a little dangerous
I would be in favor of making the origin chain and owner address required fields to avoid potential mixups like the one we had

Reply:
oky I'll make it required. I agree that it makes it less error-prone and clearer at the same time.